File: backend/internal/provider/extensions/migu.go
package extensions

import (
	"fmt"
	"net/http"
	"net/url"
	"regexp"
	"strconv"
	"strings"

	"github.com/aki-riko/Melodex/backend/internal/provider/model"
)

const (
	miguSearchBaseURL  = "http://pd.musicapp.migu.cn/MIGUM2.0/v1.0/content"
	miguContentBaseURL = "https://app.c.nf.migu.cn"
)

var (
	miguPlaylistLinkPatterns = []*regexp.Regexp{
		regexp.MustCompile(`(?i)playlistId=(\d+)`),
		regexp.MustCompile(`(?i)musicListId=(\d+)`),
		regexp.MustCompile(`(?i)/(?:playlist|songlist)/(\d+)`),
	}
	miguAlbumLinkPatterns = []*regexp.Regexp{
		regexp.MustCompile(`(?i)music\.migu\.cn/(?:v3|v5)/music/album/(\d+)`),
		regexp.MustCompile(`(?i)albumId=(\d+)`),
		regexp.MustCompile(`(?i)resourceId=(\d+)`),
	}
)

type Migu struct {
	Cookie string
}

func NewMigu(cookie string) *Migu {
	return &Migu{Cookie: strings.TrimSpace(cookie)}
}

func miguHeaders() http.Header {
	headers := make(http.Header)
	headers.Set("Referer", "http://music.migu.cn/")
	return headers
}

func (client *Migu) SearchPlaylist(keyword string) ([]model.RemoteCollection, error) {
	return client.searchPlaylists(keyword, 1, 30)
}

func (client *Migu) searchPlaylists(keyword string, page, limit int) ([]model.RemoteCollection, error) {
	query := miguSearchQuery(keyword, page, limit, `{"song":0,"album":0,"singer":0,"tagSong":0,"mvSong":0,"songlist":1,"bestShow":1}`)
	payload, err := getJSON(miguSearchBaseURL+"/search_all.do?"+query.Encode(), client.Cookie, miguHeaders())
	if err != nil {
		return nil, err
	}
	return miguPlaylists(at(payload, "songListResultData", "result")), nil
}

func (client *Migu) SearchAlbum(keyword string) ([]model.RemoteCollection, error) {
	query := miguSearchQuery(keyword, 1, 30, `{"song":0,"album":1,"singer":0,"tagSong":0,"mvSong":0,"songlist":0,"bestShow":1}`)
	payload, err := getJSON(miguSearchBaseURL+"/search_all.do?"+query.Encode(), client.Cookie, miguHeaders())
	if err != nil {
		return nil, err
	}
	albums := miguAlbums(at(payload, "albumResultData", "result"))
	if len(albums) == 0 {
		return nil, errorsForEmpty("migu albums")
	}
	return albums, nil
}

func miguSearchQuery(keyword string, page, limit int, searchSwitch string) url.Values {
	if page < 1 {
		page = 1
	}
	if limit < 1 || limit > 100 {
		limit = 30
	}
	return url.Values{
		"ua": {"Android_migu"}, "version": {"5.0.1"}, "text": {strings.TrimSpace(keyword)},
		"pageNo": {strconv.Itoa(page)}, "pageSize": {strconv.Itoa(limit)}, "searchSwitch": {searchSwitch},
	}
}

func (client *Migu) GetPlaylistSongs(id string) ([]model.Track, error) {
	return client.playlistSongs(id)
}

func (client *Migu) GetAlbumSongs(id string) ([]model.Track, error) {
	return client.albumSongs(id)
}

func (client *Migu) Playlist(id string) (*model.RemoteCollection, []model.Track, error) {
	id = strings.TrimSpace(id)
	if id == "" {
		return nil, nil, fmt.Errorf("migu playlist id is empty")
	}
	songs, err := client.playlistSongs(id)
	if err != nil {
		return nil, nil, err
	}
	playlist, infoErr := client.playlistInfo(id)
	if infoErr != nil {
		playlist = miguPlaylistFallback(id)
	}
	completeMiguPlaylistMetadata(&playlist, songs)
	return &playlist, songs, nil
}

func miguPlaylistFallback(id string) model.RemoteCollection {
	playlist := model.RemoteCollection{}
	playlist.ID, playlist.Name, playlist.Source = id, id, "migu"
	playlist.Link = miguPlaylistLink(id)
	return playlist
}

func completeMiguPlaylistMetadata(playlist *model.RemoteCollection, songs []model.Track) {
	if playlist.TrackCount == 0 {
		playlist.TrackCount = len(songs)
	}
	if playlist.Cover == "" && len(songs) != 0 {
		playlist.Cover = songs[0].Cover
	}
}

func (client *Migu) Album(id string) (*model.RemoteCollection, []model.Track, error) {
	id = strings.TrimSpace(id)
	if id == "" {
		return nil, nil, fmt.Errorf("migu album id is empty")
	}
	songs, err := client.albumSongs(id)
	if err != nil {
		return nil, nil, err
	}
	query := url.Values{"needSimple": {"00"}, "resourceType": {"2003"}, "resourceId": {id}}
	payload, err := getJSON(miguContentBaseURL+"/MIGUM2.0/v1.0/content/resourceinfo.do?"+query.Encode(), client.Cookie, miguHeaders())
	if err != nil {
		return nil, nil, err
	}
	if err := validateMiguPayload(payload, "album info"); err != nil {
		return nil, nil, err
	}
	resources := array(payload["resource"])
	if len(resources) == 0 {
		return nil, nil, errorsForEmpty("migu album info")
	}
	album := miguAlbumDetail(object(resources[0]), id)
	if album.TrackCount == 0 {
		album.TrackCount = len(songs)
	}
	for index := range songs {
		if songs[index].Album == "" {
			songs[index].Album = album.Name
		}
		if songs[index].AlbumID == "" {
			songs[index].AlbumID = album.ID
			songs[index].Extra["album_id"] = album.ID
		}
		if songs[index].Cover == "" {
			songs[index].Cover = album.Cover
		}
	}
	return &album, songs, nil
}

func (client *Migu) playlistInfo(id string) (model.RemoteCollection, error) {
	query := url.Values{"needSimple": {"00"}, "resourceType": {"2021"}, "resourceId": {id}}
	payload, err := getJSON(miguContentBaseURL+"/MIGUM2.0/v1.0/content/resourceinfo.do?"+query.Encode(), client.Cookie, miguHeaders())
	if err != nil {
		return model.RemoteCollection{}, err
	}
	if err := validateMiguPayload(payload, "playlist info"); err != nil {
		return model.RemoteCollection{}, err
	}
	resources := array(payload["resource"])
	if len(resources) == 0 {
		return model.RemoteCollection{}, errorsForEmpty("migu playlist info")
	}
	raw := object(resources[0])
	playlistID := firstText(raw, "musicListId", "id")
	if playlistID == "" {
		playlistID = id
	}
	cover := firstText(raw, "originalImgUrl")
	if cover == "" {
		cover = firstText(object(raw["imgItem"]), "img", "webpImg", "imgOri")
	}
	return model.RemoteCollection{
		ID: playlistID, Name: firstText(raw, "title", "name"), Cover: normalizeMiguImageURL(cover),
		TrackCount: integer(raw["musicNum"]), PlayCount: integer(at(raw, "opNumItem", "playNum")),
		Creator: text(raw["ownerName"]), Description: text(raw["summary"]), Source: "migu",
		Link: miguPlaylistLink(playlistID), Extra: map[string]string{"type": "playlist", "playlist_id": playlistID},
	}, nil
}

func (client *Migu) playlistSongs(id string) ([]model.Track, error) {
	return client.collectionSongs(
		miguContentBaseURL+"/MIGUM3.0/resource/playlist/song/v2.0",
		"playlistId", strings.TrimSpace(id), "migu playlist",
	)
}

func (client *Migu) albumSongs(id string) ([]model.Track, error) {
	return client.collectionSongs(
		miguContentBaseURL+"/MIGUM2.0/v1.0/content/queryAlbumSong",
		"albumId", strings.TrimSpace(id), "migu album",
	)
}

func (client *Migu) collectionSongs(endpoint, idKey, id, operation string) ([]model.Track, error) {
	if id == "" {
		return nil, fmt.Errorf("%s id is empty", operation)
	}
	const pageSize = 50
	result := make([]model.Track, 0)
	seen := make(map[string]struct{})
	for page := 1; page <= 100; page++ {
		query := url.Values{
			idKey: {id}, "pageNo": {strconv.Itoa(page)}, "pageSize": {strconv.Itoa(pageSize)},
		}
		payload, err := getJSON(endpoint+"?"+query.Encode(), client.Cookie, miguHeaders())
		if err != nil {
			return nil, err
		}
		if err := validateMiguPayload(payload, operation); err != nil {
			return nil, err
		}
		items := array(at(payload, "data", "songList"))
		for _, song := range miguSongs(items) {
			if _, ok := seen[song.ID]; ok {
				continue
			}
			seen[song.ID] = struct{}{}
			result = append(result, song)
		}
		total := integer(at(payload, "data", "totalCount"))
		if len(items) < pageSize || total > 0 && len(result) >= total {
			break
		}
	}
	if len(result) == 0 {
		return nil, errorsForEmpty(operation)
	}
	return result, nil
}

func (client *Migu) ParsePlaylist(link string) (*model.RemoteCollection, []model.Track, error) {
	link = strings.TrimSpace(link)
	for _, pattern := range miguPlaylistLinkPatterns {
		if matches := pattern.FindStringSubmatch(link); len(matches) == 2 {
			return client.Playlist(matches[1])
		}
	}
	if link != "" && !strings.Contains(link, "/") {
		return client.Playlist(link)
	}
	return nil, nil, fmt.Errorf("invalid migu playlist link")
}

func (client *Migu) ParseAlbum(link string) (*model.RemoteCollection, []model.Track, error) {
	for _, pattern := range miguAlbumLinkPatterns {
		if matches := pattern.FindStringSubmatch(strings.TrimSpace(link)); len(matches) == 2 {
			return client.Album(matches[1])
		}
	}
	return nil, nil, fmt.Errorf("invalid migu album link")
}

func (client *Migu) PlaylistCategories() ([]model.RemoteCategory, error) {
	values := []struct {
		group string
		name  string
		hot   bool
	}{
		{"语种", "华语", true}, {"语种", "欧美", true}, {"语种", "日语", false},
		{"语种", "韩语", false}, {"语种", "粤语", false}, {"风格", "流行", true},
		{"风格", "摇滚", true}, {"风格", "民谣", true}, {"风格", "电子", false},
		{"风格", "说唱", false}, {"风格", "古风", false}, {"风格", "轻音乐", false},
		{"场景", "影视", true}, {"场景", "ACG", false}, {"场景", "治愈", false},
		{"场景", "运动", false}, {"场景", "学习", false}, {"场景", "睡前", false},
	}
	result := []model.RemoteCategory{{ID: "", Name: "全部", Group: "全部", Source: "migu"}}
	for _, value := range values {
		result = append(result, model.RemoteCategory{
			ID: value.name, Name: value.name, Group: value.group, Source: "migu", Hot: value.hot,
		})
	}
	return result, nil
}

func (client *Migu) CategoryPlaylists(categoryID string, page, limit int) ([]model.RemoteCollection, error) {
	categoryID = strings.TrimSpace(categoryID)
	if categoryID == "" {
		categoryID = "华语"
	}
	playlists, err := client.searchPlaylists(categoryID, page, limit)
	if err != nil {
		return nil, err
	}
	if len(playlists) == 0 {
		return nil, errorsForEmpty("migu category playlists")
	}
	for index := range playlists {
		if playlists[index].Extra == nil {
			playlists[index].Extra = make(map[string]string)
		}
		playlists[index].Extra["category_id"] = categoryID
	}
	return playlists, nil
}

func validateMiguPayload(payload map[string]interface{}, operation string) error {
	code := text(payload["code"])
	if code != "" && code != "000000" {
		return fmt.Errorf("migu %s failed: code=%s info=%s", operation, code, text(payload["info"]))
	}
	return nil
}

func miguSongs(value interface{}) []model.Track {
	items := array(value)
	result := make([]model.Track, 0, len(items))
	for _, item := range items {
		raw := object(item)
		id := firstText(raw, "contentId", "songId", "copyrightId", "id")
		if id == "" {
			continue
		}
		name := firstText(raw, "songName", "name")
		artist := joinMiguNames(firstArray(raw, "singerList", "singers", "artists"))
		if artist == "" {
			artist = text(raw["singer"])
		}
		album := text(raw["album"])
		albumID := text(raw["albumId"])
		if albums := array(raw["albums"]); len(albums) > 0 {
			if album == "" {
				album = text(object(albums[0])["name"])
			}
			if albumID == "" {
				albumID = text(object(albums[0])["id"])
			}
		}
		size, ext := bestMiguFormat(firstArray(raw, "audioFormats", "rateFormats", "newRateFormats"))
		duration := integer(raw["duration"])
		bitrate := 0
		if size > 0 && duration > 0 {
			bitrate = int(size * 8 / 1000 / int64(duration))
		}
		cover := firstText(raw, "img1", "img2", "img3")
		if cover == "" {
			cover = pickMiguImage(firstArray(raw, "imgItems", "albumImgs"))
		}
		result = append(result, model.Track{
			ID: id, Name: name, Artist: artist, Album: album, AlbumID: albumID,
			Duration: duration, Size: size, Bitrate: bitrate, Source: "migu", Ext: ext,
			Cover: normalizeMiguImageURL(cover), Link: "https://music.migu.cn/v3/music/song/" + id,
			Extra: map[string]string{
				"content_id": text(raw["contentId"]), "copyright_id": text(raw["copyrightId"]),
				"album_id": albumID, "provider_lookup": strings.TrimSpace(name + " " + artist),
			},
		})
	}
	return result
}

func miguPlaylists(value interface{}) []model.RemoteCollection {
	items := array(value)
	result := make([]model.RemoteCollection, 0, len(items))
	for _, item := range items {
		raw := object(item)
		id := text(raw["id"])
		name := text(raw["name"])
		if id == "" || name == "" {
			continue
		}
		cover := firstText(raw, "musicListPicUrl")
		if cover == "" {
			cover = pickMiguImage(raw["imgItems"])
		}
		result = append(result, model.RemoteCollection{
			ID: id, Name: name, Cover: normalizeMiguImageURL(cover), TrackCount: integer(raw["musicNum"]),
			PlayCount: integer(raw["playNum"]), Creator: firstText(raw, "userName", "ownerName"),
			Source: "migu", Link: miguPlaylistLink(id),
			Extra: map[string]string{"type": "playlist", "playlist_id": id},
		})
	}
	return result
}

func miguAlbums(value interface{}) []model.RemoteCollection {
	items := array(value)
	result := make([]model.RemoteCollection, 0, len(items))
	for _, item := range items {
		raw := object(item)
		id := text(raw["id"])
		if id == "" {
			continue
		}
		result = append(result, model.RemoteCollection{
			ID: id, Name: text(raw["name"]), Cover: normalizeMiguImageURL(pickMiguImage(raw["imgItems"])),
			Creator: text(raw["singer"]), Description: firstText(raw, "desc", "publishDate"),
			Source: "migu", Link: miguAlbumLink(id),
			Extra: map[string]string{"type": "album", "album_id": id, "resource_type": text(raw["resourceType"])},
		})
	}
	return result
}

func miguAlbumDetail(raw map[string]interface{}, fallbackID string) model.RemoteCollection {
	id := firstText(raw, "albumId", "id")
	if id == "" {
		id = fallbackID
	}
	return model.RemoteCollection{
		ID: id, Name: text(raw["title"]), Cover: normalizeMiguImageURL(pickMiguImage(raw["imgItems"])),
		TrackCount: integer(raw["totalCount"]), PlayCount: integer(at(raw, "opNumItem", "playNum")),
		Creator: text(raw["singer"]), Description: text(raw["summary"]), Source: "migu",
		Link: miguAlbumLink(id), Extra: map[string]string{"type": "album", "album_id": id},
	}
}

func firstArray(values map[string]interface{}, keys ...string) interface{} {
	for _, key := range keys {
		if items := array(values[key]); len(items) > 0 {
			return items
		}
	}
	return nil
}

func joinMiguNames(value interface{}) string {
	names := make([]string, 0)
	for _, item := range array(value) {
		if name := text(object(item)["name"]); name != "" {
			names = append(names, name)
		}
	}
	return strings.Join(names, " / ")
}

func pickMiguImage(value interface{}) string {
	items := array(value)
	for _, preferred := range []string{"02", "01", "03"} {
		for _, item := range items {
			raw := object(item)
			if text(raw["imgSizeType"]) == preferred && text(raw["img"]) != "" {
				return text(raw["img"])
			}
		}
	}
	for _, item := range items {
		if image := firstText(object(item), "img", "webpImg", "imgOri"); image != "" {
			return image
		}
	}
	return ""
}

func bestMiguFormat(value interface{}) (int64, string) {
	var bestSize int64
	ext := "mp3"
	for _, item := range array(value) {
		raw := object(item)
		size := firstInt64(raw, "asize", "androidSize", "size", "isize")
		if size <= bestSize {
			continue
		}
		bestSize = size
		formatType := strings.ToUpper(text(raw["formatType"]))
		formatCode := firstText(raw, "aformat", "androidFileType", "fileType", "iformat")
		if strings.Contains(formatType, "SQ") || strings.HasPrefix(formatCode, "011") {
			ext = "flac"
		} else {
			ext = "mp3"
		}
	}
	return bestSize, ext
}

func normalizeMiguImageURL(value string) string {
	value = strings.TrimSpace(value)
	if strings.HasPrefix(value, "//") {
		return "https:" + value
	}
	if strings.HasPrefix(value, "/") {
		return "https://d.musicapp.migu.cn" + value
	}
	return value
}

func miguPlaylistLink(id string) string {
	return "https://music.migu.cn/v5/#/playlist?playlistId=" + id + "&playlistType=ordinary"
}

func miguAlbumLink(id string) string {
	return "https://music.migu.cn/v3/music/album/" + id
}

File: backend/provider_bridge/platforms/migu.py
"""Migu collection adapter hosted by the provider sidecar."""

from __future__ import annotations

import re
from urllib.parse import urlencode

from provider_bridge.collection_common import (
    at,
    category_payload,
    collection_payload,
    first,
    first_integer,
    integer,
    join_names,
    list_value,
    object_value,
    page_limit,
    regex_link_id,
    string,
    track_payload,
)
from provider_bridge.platform_http import PlatformHTTP


SEARCH_URL = "http://pd.musicapp.migu.cn/MIGUM2.0/v1.0/content"
CONTENT_URL = "https://app.c.nf.migu.cn"
HEADERS = {"Referer": "http://music.migu.cn/"}


class MiguCollections:
    source = "migu"

    def __init__(self, cookie: str = "", session=None):
        self.http = PlatformHTTP(cookie, session)

    def search_playlist(self, keyword: str, **_):
        payload = self.http.get(SEARCH_URL + "/search_all.do?" + urlencode(self._search_query(keyword, 1, 30, True)), headers=HEADERS)
        return self._playlists(at(payload, "songListResultData", "result"))

    def search_album(self, keyword: str, **_):
        payload = self.http.get(SEARCH_URL + "/search_all.do?" + urlencode(self._search_query(keyword, 1, 30, False)), headers=HEADERS)
        albums = self._albums(at(payload, "albumResultData", "result"))
        if not albums:
            raise ValueError("migu albums response is empty")
        return albums

    @staticmethod
    def _search_query(keyword, page, limit, playlist):
        switches = (
            '{"song":0,"album":0,"singer":0,"tagSong":0,"mvSong":0,"songlist":1,"bestShow":1}'
            if playlist else
            '{"song":0,"album":1,"singer":0,"tagSong":0,"mvSong":0,"songlist":0,"bestShow":1}'
        )
        return {"ua": "Android_migu", "version": "5.0.1", "text": string(keyword), "pageNo": max(1, integer(page)), "pageSize": min(max(integer(limit) or 30, 1), 100), "searchSwitch": switches}

    def playlist(self, identifier: str, **_):
        identifier = string(identifier)
        if not identifier:
            raise ValueError("migu playlist id is empty")
        songs = self._collection_songs("/MIGUM3.0/resource/playlist/song/v2.0", "playlistId", identifier, "migu playlist")
        try:
            playlist = self._playlist_info(identifier)
        except Exception:
            playlist = collection_payload(identifier=identifier, source=self.source, name=identifier, link=self._playlist_link(identifier))
        playlist["track_count"] = playlist["track_count"] or len(songs)
        if not playlist["cover"] and songs:
            playlist["cover"] = songs[0]["cover"]
        return playlist, songs

    def album(self, identifier: str, **_):
        identifier = string(identifier)
        if not identifier:
            raise ValueError("migu album id is empty")
        songs = self._collection_songs("/MIGUM2.0/v1.0/content/queryAlbumSong", "albumId", identifier, "migu album")
        query = urlencode({"needSimple": "00", "resourceType": 2003, "resourceId": identifier})
        payload = self.http.get(CONTENT_URL + "/MIGUM2.0/v1.0/content/resourceinfo.do?" + query, headers=HEADERS)
        self._validate(payload, "album info")
        resources = list_value(payload.get("resource"))
        if not resources:
            raise ValueError("migu album info response is empty")
        album = self._album_detail(object_value(resources[0]), identifier)
        album["track_count"] = album["track_count"] or len(songs)
        for song in songs:
            song["album"] = song["album"] or album["name"]
            song["album_id"] = song["album_id"] or album["id"]
            song["extra"]["album_id"] = song["album_id"]
            song["cover"] = song["cover"] or album["cover"]
        return album, songs

    def parse_playlist(self, link: str, **_):
        identifier = regex_link_id(link, (r"playlistId=(\d+)", r"musicListId=(\d+)", r"/(?:playlist|songlist)/(\d+)"))
        if not identifier and string(link) and "/" not in string(link):
            identifier = string(link)
        if not identifier:
            raise ValueError("invalid migu playlist link")
        return self.playlist(identifier)

    def parse_album(self, link: str, **_):
        identifier = regex_link_id(link, (r"music\.migu\.cn/(?:v3|v5)/music/album/(\d+)", r"albumId=(\d+)", r"resourceId=(\d+)"))
        if not identifier:
            raise ValueError("invalid migu album link")
        return self.album(identifier)

    def recommend(self, **_):
        return self.search_playlist("华语")

    def categories(self, **_):
        values = (
            ("语种", "华语", True), ("语种", "欧美", True), ("语种", "日语", False), ("语种", "韩语", False),
            ("语种", "粤语", False), ("风格", "流行", True), ("风格", "摇滚", True), ("风格", "民谣", True),
            ("风格", "电子", False), ("风格", "说唱", False), ("风格", "古风", False), ("风格", "轻音乐", False),
            ("场景", "影视", True), ("场景", "ACG", False), ("场景", "治愈", False), ("场景", "运动", False),
            ("场景", "学习", False), ("场景", "睡前", False),
        )
        result = [category_payload(identifier="", source=self.source, name="全部", group="全部")]
        result.extend(category_payload(identifier=name, source=self.source, name=name, group=group, hot=hot) for group, name, hot in values)
        return result

    def category(self, category_id: str, page: int = 1, limit: int = 30, **_):
        category_id = string(category_id) or "华语"
        playlists = self._search_playlists(category_id, page, limit)
        if not playlists:
            raise ValueError("migu category playlists response is empty")
        for playlist in playlists:
            playlist["extra"]["category_id"] = category_id
        return playlists

    def user_playlists(self, **_):
        raise ValueError("migu user playlists are not supported")

    def _search_playlists(self, keyword, page, limit):
        payload = self.http.get(SEARCH_URL + "/search_all.do?" + urlencode(self._search_query(keyword, page, limit, True)), headers=HEADERS)
        return self._playlists(at(payload, "songListResultData", "result"))

    def _playlist_info(self, identifier):
        query = urlencode({"needSimple": "00", "resourceType": 2021, "resourceId": identifier})
        payload = self.http.get(CONTENT_URL + "/MIGUM2.0/v1.0/content/resourceinfo.do?" + query, headers=HEADERS)
        self._validate(payload, "playlist info")
        resources = list_value(payload.get("resource"))
        if not resources:
            raise ValueError("migu playlist info response is empty")
        raw = object_value(resources[0])
        playlist_id = first(raw, "musicListId", "id") or identifier
        image = first(raw, "originalImgUrl") or first(object_value(raw.get("imgItem")), "img", "webpImg", "imgOri")
        return collection_payload(
            identifier=playlist_id, source=self.source, name=first(raw, "title", "name"), cover=self._image(image),
            track_count=integer(raw.get("musicNum")), play_count=integer(at(raw, "opNumItem", "playNum")),
            creator=string(raw.get("ownerName")), description=string(raw.get("summary")), link=self._playlist_link(playlist_id),
            extra={"type": "playlist", "playlist_id": playlist_id},
        )

    def _collection_songs(self, endpoint, key, identifier, operation):
        result, seen = [], set()
        for page in range(1, 101):
            query = urlencode({key: identifier, "pageNo": page, "pageSize": 50})
            payload = self.http.get(CONTENT_URL + endpoint + "?" + query, headers=HEADERS)
            self._validate(payload, operation)
            items = list_value(at(payload, "data", "songList"))
            for song in self._songs(items):
                if song["id"] not in seen:
                    seen.add(song["id"])
                    result.append(song)
            total = integer(at(payload, "data", "totalCount"))
            if len(items) < 50 or (total and len(result) >= total):
                break
        if not result:
            raise ValueError(operation + " response is empty")
        return result

    @staticmethod
    def _validate(payload, operation):
        code = string(payload.get("code"))
        if code and code != "000000":
            raise ValueError(f"migu {operation} failed: code={code}")

    def _songs(self, value):
        result = []
        for raw_value in list_value(value):
            raw = object_value(raw_value)
            identifier = first(raw, "contentId", "songId", "copyrightId", "id")
            if not identifier:
                continue
            name = first(raw, "songName", "name")
            artist = join_names(raw.get("singerList") or raw.get("singers") or raw.get("artists"), separator=" / ") or string(raw.get("singer"))
            album = string(raw.get("album"))
            album_id = string(raw.get("albumId"))
            albums = list_value(raw.get("albums"))
            if albums:
                album = album or string(object_value(albums[0]).get("name"))
                album_id = album_id or string(object_value(albums[0]).get("id"))
            size, ext = self._best_format(raw.get("audioFormats") or raw.get("rateFormats") or raw.get("newRateFormats"))
            duration = integer(raw.get("duration"))
            bitrate = size * 8 // 1000 // duration if size and duration else 0
            image = first(raw, "img1", "img2", "img3") or self._pick_image(raw.get("imgItems") or raw.get("albumImgs"))
            result.append(track_payload(
                identifier=identifier, source=self.source, name=name, artist=artist, album=album, album_id=album_id,
                duration=duration, size=size, bitrate=bitrate, ext=ext, cover=self._image(image),
                link="https://music.migu.cn/v3/music/song/" + identifier,
                extra={"content_id": string(raw.get("contentId")), "copyright_id": string(raw.get("copyrightId")), "album_id": album_id, "provider_lookup": f"{name} {artist}".strip()},
            ))
        return result

    def _playlists(self, value):
        result = []
        for raw_value in list_value(value):
            raw = object_value(raw_value)
            identifier, name = string(raw.get("id")), string(raw.get("name"))
            if identifier and name:
                image = first(raw, "musicListPicUrl") or self._pick_image(raw.get("imgItems"))
                result.append(collection_payload(
                    identifier=identifier, source=self.source, name=name, cover=self._image(image),
                    track_count=integer(raw.get("musicNum")), play_count=integer(raw.get("playNum")),
                    creator=first(raw, "userName", "ownerName"), link=self._playlist_link(identifier),
                    extra={"type": "playlist", "playlist_id": identifier},
                ))
        return result

    def _albums(self, value):
        result = []
        for raw_value in list_value(value):
            raw = object_value(raw_value)
            identifier = string(raw.get("id"))
            if identifier:
                result.append(collection_payload(
                    identifier=identifier, source=self.source, name=string(raw.get("name")),
                    cover=self._image(self._pick_image(raw.get("imgItems"))), creator=string(raw.get("singer")),
                    description=first(raw, "desc", "publishDate"), link=self._album_link(identifier),
                    extra={"type": "album", "album_id": identifier, "resource_type": string(raw.get("resourceType"))},
                ))
        return result

    def _album_detail(self, raw, fallback_id):
        identifier = first(raw, "albumId", "id") or fallback_id
        return collection_payload(
            identifier=identifier, source=self.source, name=string(raw.get("title")),
            cover=self._image(self._pick_image(raw.get("imgItems"))), track_count=integer(raw.get("totalCount")),
            play_count=integer(at(raw, "opNumItem", "playNum")), creator=string(raw.get("singer")),
            description=string(raw.get("summary")), link=self._album_link(identifier),
            extra={"type": "album", "album_id": identifier},
        )

    @staticmethod
    def _pick_image(value):
        items = list_value(value)
        for preferred in ("02", "01", "03"):
            for item in items:
                raw = object_value(item)
                if string(raw.get("imgSizeType")) == preferred and string(raw.get("img")):
                    return string(raw.get("img"))
        for item in items:
            raw = object_value(item)
            image = first(raw, "img", "webpImg", "imgOri")
            if image:
                return image
        return ""

    @staticmethod
    def _best_format(value):
        best_size, extension = 0, "mp3"
        for item in list_value(value):
            raw = object_value(item)
            size = 0
            for key in ("asize", "androidSize", "size", "isize"):
                size = integer(raw.get(key))
                if size:
                    break
            if size <= best_size:
                continue
            best_size = size
            fmt = string(raw.get("formatType")).upper()
            code = first(raw, "aformat", "androidFileType", "fileType", "iformat")
            extension = "flac" if "SQ" in fmt or code.startswith("011") else "mp3"
        return best_size, extension

    @staticmethod
    def _image(value):
        value = string(value)
        if value.startswith("//"):
            return "https:" + value
        if value.startswith("/"):
            return "https://d.musicapp.migu.cn" + value
        return value

    @staticmethod
    def _playlist_link(identifier):
        return "https://music.migu.cn/v5/#/playlist?playlistId=" + identifier + "&playlistType=ordinary"

    @staticmethod
    def _album_link(identifier):
        return "https://music.migu.cn/v3/music/album/" + identifier
